Data Instruction support

wynick27 opened the pull request:
Implemented feature mentioned in #632 #716
Currently supports disassemble and assemble integer (1-16 bytes), float (single and double), string (ascii and unicode) and mmx, xmm, ymm data instructions.
80 bit extended float is not supported.
Adds another analysis that can analyze functions, xrefs and data. (analadv command)
Allows code mixed with data.
Currently information is stored in encode map file.
![image](https://cloud.githubusercontent.com/assets/6976053/16130656/460712fc-33d8-11e6-998a-2a02d99bd84b.png)


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/x64dbg/x64dbg/758)
<!-- Reviewable:end -->

mrexodia commented on src/dbg/_exports.cpp:
```diff
@@ -1049,6 +1056,41 @@ extern "C" DLL_EXPORT duint _dbg_sendmessage(DBGMSG type, void* param1, void* pa
+    {
+        EncodeMapReleaseBuffer(param1);
+    }
+
```
this will fall through to `DBG_GET_STRING_AT` right?

mrexodia commented on src/dbg/advancedanalysis.cpp:
```diff
@@ -0,0 +1,373 @@
+
+void AdvancedAnalysis::analyzeFunction(duint entryPoint, bool writedata)
+{
+    //BFS through the disassembly starting at entryPoint
```
Notice that this code does not handle overlapping blocks correctly, not sure if this will cause problems later.

wynick27 replied:
I know, maybe I'll resolve overlapping problem later. There could be gaps between basic blocks, especially in control flow obfuscated executable.

mrexodia commented on src/dbg/advancedanalysis.cpp:
```diff
@@ -0,0 +1,373 @@
+    }
+}
+
+bool isFloatInstruction(x86_insn opcode)
```
Maybe we should add this to `capstone_wrapper`? Also this list appears to be pretty incomplete, is this just for float references?

wynick27 replied:
Yes, it's a work in progress, I plan to add simd instruction as well, but it takes a lot of time to look up the manuals.

mrexodia commented on src/dbg/assemble.cpp:
```diff
@@ -181,8 +191,19 @@ bool assembleat(duint addr, const char* instruction, int* size, char* error, boo
+    {
+        if(destSize > 16)
+        {
+            newbuffer = new unsigned char[destSize];  //retry for long string sequence
```
We have `Memory<unsigned char*>` for this to prevent possible memory leaks.

mrexodia commented on src/dbg/breakpoint.cpp:
```diff
@@ -97,7 +97,8 @@ bool BpNew(duint Address, bool Enable, bool Singleshot, short OldBytes, BP_TYPE
 bool BpGet(duint Address, BP_TYPE Type, const char* Name, BREAKPOINT* Bp)
 {
-    ASSERT_DEBUGGING("Export call");
+    if(!DbgIsDebugging())
```
Why is this here? Is BpGet called when not debugging? The assert should take care of that (although I never run in debug mode).

wynick27 replied:
BpGet  and PatchEnum always get galled when debugging is being stopped.

mrexodia commented:
It's looking good so far, but I have to check the code better so I will merge it later.

wynick27 commented:
I fixed the bug in dbg message function and uses Memory template to allocate memory.
I also added temp code list support for encode map, addresses in the list will always be decoded as code and any other address decoding that conflicts with it will be decoded as data.
Currently CIP is added to the list and if instruction in CIP is branch its target is also added.
These screenshots shows the effect of code list.
I write this test program with junk code at rva 0x1a, without analysis, it's wrongly decoded as instruction.
![image](https://cloud.githubusercontent.com/assets/6976053/16157436/8d0228b4-3486-11e6-9705-e7757422d3bf.png)
After CIP points to that address, it's correctly decoded.
![image](https://cloud.githubusercontent.com/assets/6976053/16157479/b6fcab26-3486-11e6-8ef1-6d9ebfe8fb53.png)

mrexodia commented:
Thanks a lot, I will test the code and commit fixes if necessary...
